File: attribouter/src/main/java/me/jfenn/attribouter/wedges/ContributorWedge.java
package me.jfenn.attribouter.wedges;

import android.content.Context;
import android.content.res.XmlResourceParser;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;

import androidx.annotation.Nullable;
import me.jfenn.attribouter.R;
import me.jfenn.attribouter.data.github.GitHubData;
import me.jfenn.attribouter.data.github.UserData;
import me.jfenn.attribouter.dialogs.UserDialog;
import me.jfenn.attribouter.interfaces.Mergeable;
import me.jfenn.attribouter.utils.ResourceUtils;
import me.jfenn.attribouter.wedges.link.EmailLinkWedge;
import me.jfenn.attribouter.wedges.link.GitHubLinkWedge;
import me.jfenn.attribouter.wedges.link.LinkWedge;
import me.jfenn.attribouter.wedges.link.WebsiteLinkWedge;

public class ContributorWedge extends Wedge<ContributorWedge.ViewHolder> implements Mergeable<ContributorWedge> {

    @Nullable
    public String login;
    @Nullable
    public String name;
    @Nullable
    public String avatarUrl;
    @Nullable
    public String bio;
    @Nullable
    public String blog;
    @Nullable
    public String email;
    @Nullable
    Integer position;
    @Nullable
    public String task;

    private boolean isHidden;

    public ContributorWedge(XmlResourceParser parser) throws XmlPullParserException, IOException {
        this(parser.getAttributeValue(null, "login"),
                parser.getAttributeValue(null, "name"),
                parser.getAttributeValue(null, "avatar"),
                parser.getAttributeValue(null, "task"),
                parser.getAttributeIntValue(null, "position", -1),
                parser.getAttributeValue(null, "bio"),
                parser.getAttributeValue(null, "blog"),
                parser.getAttributeValue(null, "email"));

        isHidden = parser.getAttributeBooleanValue(null, "hidden", false);
        addChildren(parser);
    }

    protected ContributorWedge(@Nullable String login, @Nullable String name, @Nullable String avatarUrl, @Nullable String task, @Nullable Integer position, @Nullable String bio, @Nullable String blog, @Nullable String email) {
        super(R.layout.item_attribouter_contributor);
        this.login = login;
        this.name = name;
        this.avatarUrl = avatarUrl;
        this.task = task;
        this.position = position != null && position >= 0 ? position : null;
        this.bio = bio;
        this.blog = blog;
        this.email = email;

        if (login != null)
            addChild(new GitHubLinkWedge(login, 1));
        if (blog != null)
            addChild(new WebsiteLinkWedge(blog, 2));
        if (email != null)
            addChild(new EmailLinkWedge(email, -1));

        if (login != null && !hasAll())
            addRequest(new UserData(login));
    }

    @Override
    public void onInit(GitHubData data) {
        if (data instanceof UserData) {
            UserData user = (UserData) data;
            merge(new ContributorWedge(
                    user.login,
                    user.name,
                    user.avatar_url,
                    task == null ? "Contributor" : null,
                    null,
                    user.bio,
                    user.blog,
                    user.email
            ));
        }
    }

    @Nullable
    public String getName() {
        return name != null ? name : login;
    }

    public ContributorWedge merge(ContributorWedge contributor) {
        if ((name == null || !name.startsWith("^")) && contributor.name != null)
            name = contributor.name;
        if ((avatarUrl == null || !avatarUrl.startsWith("^")) && contributor.avatarUrl != null)
            avatarUrl = contributor.avatarUrl;
        if ((bio == null || !bio.startsWith("^")) && contributor.bio != null && !contributor.bio.isEmpty())
            bio = contributor.bio;
        if ((blog == null || !blog.startsWith("^")) && contributor.blog != null && !contributor.blog.isEmpty())
            blog = contributor.blog;
        if ((email == null || !email.startsWith("^")) && contributor.email != null && !contributor.email.isEmpty())
            email = contributor.email;
        if ((task == null || !task.startsWith("^")) && contributor.task != null)
            task = contributor.task;

        for (Wedge child : contributor.getChildren())
            addChild(child);

        return this;
    }

    @Override
    public boolean hasAll() {
        return name != null && name.startsWith("^") && bio != null && bio.startsWith("^") && blog != null && blog.startsWith("^") && email != null && email.startsWith("^");
    }

    @Override
    public boolean isHidden() {
        return isHidden;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof ContributorWedge) {
            ContributorWedge contributor = (ContributorWedge) obj;
            return (login != null && contributor.login != null && login.toLowerCase().equals(contributor.login.toLowerCase())) || super.equals(obj);
        } else return super.equals(obj);
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    @Override
    public void bind(Context context, ViewHolder viewHolder) {
        ResourceUtils.setImage(context, avatarUrl, viewHolder.imageView);
        viewHolder.nameView.setText(ResourceUtils.getString(context, getName()));
        if (task != null) {
            viewHolder.taskView.setVisibility(View.VISIBLE);
            viewHolder.taskView.setText(ResourceUtils.getString(context, task));
        } else viewHolder.taskView.setVisibility(View.GONE);

        if (ResourceUtils.getString(context, bio) != null) {
            viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    new UserDialog(view.getContext(), ContributorWedge.this)
                            .show();
                }
            });
        } else {
            LinkWedge importantLink = null;
            View.OnClickListener clickListener = null;
            for (LinkWedge link : getChildren(LinkWedge.class)) {
                if (!link.isHidden() && (importantLink == null || link.getPriority() > importantLink.getPriority())) {
                    View.OnClickListener listener = link.getListener(context);
                    if (listener != null) {
                        importantLink = link;
                        clickListener = listener;
                    }
                }
            }

            viewHolder.itemView.setOnClickListener(clickListener);
        }
    }

    protected static class ViewHolder extends Wedge.ViewHolder {

        protected ImageView imageView;
        protected TextView nameView;
        protected TextView taskView;

        protected ViewHolder(View v) {
            super(v);
            imageView = v.findViewById(R.id.image);
            nameView = v.findViewById(R.id.name);
            taskView = v.findViewById(R.id.task);
        }
    }
}

File: attribouter/src/main/java/me/jfenn/attribouter/wedges/ContributorWedge.kt
package me.jfenn.attribouter.wedges

import android.content.Context
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import me.jfenn.attribouter.R
import me.jfenn.attribouter.dialogs.UserDialog
import me.jfenn.attribouter.utils.ResourceUtils
import me.jfenn.attribouter.utils.equalsProvider
import me.jfenn.attribouter.utils.loadDrawable
import me.jfenn.attribouter.utils.toTitleString
import me.jfenn.gitrest.model.ProviderString
import me.jfenn.gitrest.model.User

open class ContributorWedge(
        login: String? = null,
        name: String? = null,
        avatarUrl: String? = null,
        profileUrl: String? = null,
        websiteUrl: String? = null,
        task: String? = "Contributor",
        position: Int = 0,
        bio: String? = null,
        email: String? = null
) : Wedge<ContributorWedge.ViewHolder>(R.layout.attribouter_item_contributor), Comparable<ContributorWedge> {

    var login: String? by attr("login", login)
    var name: String? by attr("name", name)
    var avatar: String? by attr("avatar", avatarUrl)
    var profileUrl: String? by attr("profileUrl", profileUrl)
    var websiteUrl: String? by attr("websiteUrl", websiteUrl)
    var task: String? by attr("task", task)
    var position: Int by object : attr<ContributorWedge, Int>("position", position) {
        override fun apply(original: Int?, value: Int?): Int? = if (value != null && value != 0) value else original
    }
    var bio: String? by attr("bio", bio)
    var email: String? by attr("email", email)
    override var isHidden: Boolean by attr("hidden", false)

    override fun onCreate() {
        initChildren()

        login?.let {
            lifecycle?.launch {
                withContext(Dispatchers.IO) {
                    lifecycle?.client?.getUser(it)
                }.let { user ->
                    user?.let { onContributor(user) }
                }
            }
        }
    }

    fun initChildren() {
        // try to guess profile URL from id
        (profileUrl ?: login?.let { ProviderString(it).inferUrl() })?.let { url ->
            login?.let { userId ->
                val id = ProviderString(userId)
                addChild(ProfileLinkWedge(
                        name = id.provider.toTitleString(),
                        url = url,
                        icon = "@drawable/attribouter_ic_${id.provider}",
                        priority = 0
                ).create(lifecycle))
            } ?: addChild(ProfileLinkWedge(url = url, priority = 1).create(lifecycle))
        }

        email?.let { addChild(EmailLinkWedge(it, 0)) }
        websiteUrl?.let { addChild(WebsiteLinkWedge(it, 2)) }
    }

    open fun onContributor(data: User) {
        login = data.providerString.toString()
        name = data.name
        avatar = data.avatarUrl
        profileUrl = data.url
        websiteUrl = data.websiteUrl
        bio = data.bio
        email = data.email

        initChildren()
        notifyItemChanged()
    }

    fun getCanonicalName(): String? {
        return name ?: login
    }

    override fun equals(other: Any?): Boolean {
        return (other as? ContributorWedge)?.let {
            login.equalsProvider(other.login)
        } ?: super.equals(other)
    }

    override fun compareTo(other: ContributorWedge): Int {
        if (position == 0)
            return 1;
        if (other.position == 0)
            return -1;
        return position - other.position
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    override fun bind(context: Context, viewHolder: ViewHolder) {
        viewHolder.imageView?.apply {
            context.loadDrawable(avatar, R.drawable.attribouter_image_avatar) {
                setImageDrawable(it)
            }
        }

        viewHolder.nameView?.apply {
            text = ResourceUtils.getString(context, getCanonicalName())
        }

        viewHolder.taskView?.apply {
            task?.let {
                text = ResourceUtils.getString(context, it)
                visibility = View.VISIBLE
            } ?: run {
                visibility = View.GONE
            }
        }

        val links = getTypedChildren<LinkWedge>().filter { !it.isHidden }.sorted()

        viewHolder.firstLinkView?.apply {
            links.getOrNull(0)?.let { link ->
                link.bind(context, LinkWedge.ViewHolder(this))
                visibility = View.VISIBLE
            } ?: run {
                visibility = View.GONE
            }
        }

        viewHolder.secondLinkView?.apply {
            links.getOrNull(1)?.let { link ->
                link.bind(context, LinkWedge.ViewHolder(this))
                visibility = View.VISIBLE
            } ?: run {
                visibility = View.GONE
            }
        }

        viewHolder.itemView.apply {
            if (ResourceUtils.getString(context, bio) != null) {
                setOnClickListener {
                    UserDialog(context, this@ContributorWedge).show()
                }
            } else {
                var importantLink: LinkWedge? = null
                var clickListener: View.OnClickListener? = null
                for (link in links) {
                    if (importantLink == null || link.priority > importantLink.priority) {
                        link.getListener(context)?.let {
                            importantLink = link
                            clickListener = it
                        }
                    }
                }

                setOnClickListener(clickListener)
            }
        }
    }

    open class ViewHolder(v: View) : Wedge.ViewHolder(v) {
        var imageView: ImageView? = v.findViewById(R.id.image)
        var nameView: TextView? = v.findViewById(R.id.name)
        var taskView: TextView? = v.findViewById(R.id.task)

        var firstLinkView: View? = v.findViewById(R.id.link_1)
        var secondLinkView: View? = v.findViewById(R.id.link_2)
    }
}
